accepted/future-releases/unified-collections/feature-specification.md

```diff
@@ -0,0 +1,802 @@
+kind of literal).
+
+We require that at least one component unambiguously determine the literal form,
+otherwise it is an error. So, given:
+
+```dart
+bool b = true;
+dynamic x = <int, int>{};
+Iterable l = [];
+Map m = {};
+```
+
+Then:
+
+```dart
+{...x} // An error, because it is ambiguous.
+{...x, ...l} // Statically resolved to a set, runtime error on the spread.
+{...x, ...m} // Statically resolved to a map, no runtime error.
+{...l, ...m} // Static error, because it must be both a set and a map.
+```
+
+The algorithm can be modified to catch this error eagerly by tracking four
+result states for element inference:
+
+*   `MapElement<K, V>` – The element constrains the literal to a map with key
+    type `K` and element type `V`.
+
+*   `SetElement<T>` – The element constrains the literal to a set with element
+    type `T`.
+
+*   `DynamicElement` – The element does not constrain the literal to a set or
+    map, but constrains the element/key/value type to `dynamic`.
+
+*   `ErrorElement` – The element is erroneous.
+
+Alternatively, error checking for invalid spreads can be done separately.
+
+In `setOrMapLiteral`, the inferred type of an `element` is a set element type
+`T`, a pair of a key type `K` and a value type `V`, or both. It is computed
+relative to a context type `P`:
+
+*   If `element` is an `expressionElement` with expression `e1`:
+
+    *   If `P` is `?` then the inferred set element type of `element` is the
+        inferred type of the expression `e1` in context `?`.
+
+    *   If `P` is `Set<Ps>` then the inferred set element type of `element` is
+        the inferred type of the expression `e1` in context `Ps`.
+
+        **TODO: what if `P` is `Iterable<R>`? Is this allowed?:**
+
+        ```dart
+        Iterable<int> = {123 as dynamic}; // Infers Set<int>?
+        ```
+
+*   If `element` is a `mapEntry` `ek: ev`:
+
+    *   If `P` is `?` then the inferred key type of `element` is the inferred
+        type of `ek` in context `?` and the inferred value type of `element` is
+        the inferred type of `ev` in context `?`.
+
+    *   If `P` is `Map<Pk, Pv>` then the inferred key type of `element` is the
+        inferred type of `ek` in context `Pk` and the inferred value type of `element`
+        is the inferred type of `ev` in context `Pv`.
+
+*   If `element` is a `spreadElement` with expression `e1`:
+
+    *   If `P` is `?` then let `S` be the inferred type of `e1` in context `?`:
+
+        *   If `S` is a subtype of `Iterable<Object>` and not a subtype of
+            `Map<Object, Object>`, then the inferred set element type of
+            `element` is `T` where `T` is the type such that `Iterable<T>` is a
+            superinterface of `S` (the result of constraint matching for `X`
+            using the constraint `S <: Iterable<X>`).
+
+        *   If `S` is a subtype of `Map<Object, Object>` and not a subtype of
+            `Iterable<Object>`, then the inferred key type of `element` is `K`
```

Comment:
My text does have this requirement.  I'm inclined to keep the requirement.  We could make it default to map, but that may or may not be the right thing.  If it's ambiguous, they can always put a type argument on the literal and this will just work (since we don't have the negative check on the case where there is a downwards context).

> With an explicit context type `Set<P>` we could even allow it to be used as a set.

This already works.

Reply:
We talked about this again.

The case that is not covered is a non-`Null` type that is a subtype of both `Map<Object, Object>` and `Iterable<Object>`.
By default it's an error.

We could, and I think should, make it work as both. Basically:

> * If `S` is a non-`Null` subtype of `Iterable<Object>`, then the inferred element type of `element` is `is `T` where `T` is the type such that `Iterable<T>` is a superinterface of `S` (the result of constraint matching for `X` using the constraint `S <: Iterable<X>`).
> * If `S` is a non-`Null` subtype of `Map<Object, Object>`, 
then the inferred key type of `element` is `K`and the inferred value type of `element` is `V`, where `K` and `V` are the types such that `Map<K, V>` is a superinterface of `S` (the result of constraint matching for `X` and `Y` using the constraint `S <: Map<X, Y>`). *(`S` may be a sub-type of both `Iterable<Object>` and `Map<Object, Object>`, in which we produce both a set element type here, and a key/value pair here and rely on other elements to disambiguate.)*


This does introduce a case where the same element expression has both types of inferred types, and they are not dynamic.
It's technically possible (no matter how ridiculous) to have a class implementing `Iterable<String>` and `Map<int, double>`.

It would allow us to do `{1, ...x}` where `x` implements both `Iterable` and `Map`. If some API starts returning a subtype of `Map` that also implements `Iterable`, instead of just `Map`, it would be sad that we broke existing correct code.
We would have to give up on `{...x}` though.

Reply:
Agreed.  This works out fine, I think - it just adds the possibility of a branch returning both a set element type and a key/value type pair, where the types are not just dynamic.